retworkx/custom_return_types.pyi

```diff
@@ -0,0 +1,60 @@
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# This file contains only type annotations for PyO3 functions and classes
+# For implementation details, see __init__.py and lib.rs
+
+from typing import Any, Generic, List, Iterable, Mapping, TypeVar, Tuple, overload
+from collections.abc import ABC, Sequence
+from typing_extensions import Self
+
+S = TypeVar("S")
+T_co = TypeVar("T_co", covariant=True)
+
+class RetworkxCustomVecIter(Generic[T_co], Sequence[T_co], ABC):
+    def __init__(self) -> None: ...
+    def __eq__(self, other: Sequence[T_co]) -> bool: ...
+    @overload
+    def __getitem__(self, index: int) -> T_co: ...
+    @overload
+    def __getitem__(self: Self, index: slice) -> Self: ...
+    def __getstate__(self) -> Any: ...
+    def __hash__(self) -> int: ...
+    def __len__(self) -> int: ...
+    def __ne__(self, other: Sequence[T_co]) -> bool: ...
+    def __setstate__(self, state) -> None: ...
+
+class RetworkxCustomHashMapIter(Generic[S, T_co], Mapping[S, T_co], ABC):
+    def __init__(self) -> None: ...
+    def items(self) -> Iterable[Tuple[S, T_co]]: ...
+    def keys(self) -> Iterable[S]: ...
+    def values(self) -> Iterable[T_co]: ...
+    def __contains__(self, other: S) -> bool: ...
+    def __eq__(self, other: Mapping[S, T_co]) -> bool: ...
+    def __getitem__(self, index: S) -> T_co: ...
+    def __getstate__(self) -> Any: ...
+    def __hash__(self) -> int: ...
+    def __iter__(self) -> Iterable[S]: ...
```

Comment:
I missed this first time - technically `__iter__` returns an `Iterator`, not an `Iterable` because it starts the iteration.  All the other `Iterable` hints are correct, though.

Reply:
Actually `__iter__` and `keys` return the same type in Rust so I'm thinking if we should use `Iterator` in all places.

Reply:
If that's the case, it's possibly a bug - it's a break with the normal Python API.  If I use `collections.abc` to build up a mapping-like object from the minimal set of abstract methods, we get:
```python
In [1]: import collections.abc
   ...:
   ...: class A(collections.abc.Mapping):
   ...:     def __init__(self, d):
   ...:         self.d = d
   ...:     def __getitem__(self, k):
   ...:         return self.d[k]
   ...:     def __iter__(self):
   ...:         return iter(self.d)
   ...:     def __len__(self):
   ...:         return len(self.d)
   ...:
   ...: next(A({}).keys())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-63ecf9188336> in <module>
     11         return len(self.d)
     12
---> 13 next(A({}).keys())

TypeError: 'KeysView' object is not an iterator
```

I would leave the others as `Iterable`, because if your Rust/Python conversion changes in the future, `Iterable` will still fulfill the requirements.  `Iterator` fulfills everything needed for `Iterable`, and I assume it's covariant in its argument, so it's ok for now, but you may as well be as restrictive as you can be when defining your output types.

Reply:
After almost half a year, some feedback: mypy agrees with Jake, I changed the code

Reply:
Haha, I think Georgios was also correct at the time he wrote the comment. Maybe PyO3 changed their default conversions in the interim? Or mypy just got stricter about magic methods.